src/server.ts

```diff
@@ -212,234 +223,256 @@ export class SubscriptionServer {
+      });
+      this.sendMessage(connectionContext, opId, MessageTypes.GQL_COMPLETE, null);
+      nextTick(() => {
+        this.processQueue(connectionContext, opId);
+      });
+    } catch (error) {
+      operationContext.processingStart = false;
+      nextTick(() => {
+        this.processQueue(connectionContext, opId);
+      });
+      if (error.errors) {
+        this.sendMessage(connectionContext, opId, MessageTypes.GQL_DATA, {
+          errors: error.errors,
+        });
+      } else {
+        this.sendError(connectionContext, opId, { message: error.message });
+      }
+      this.unsubscribe(connectionContext, opId);
+    }
+  }
+
+  private processQueue(connectionContext: ConnectionContext, opId: string) {
+    const operationContext = connectionContext.operations.get(opId);
+    if (!operationContext || operationContext.processingStart) {
+      return;
+    }
+    const message = operationContext.messageQueue.shift();
+    if (!message) {
+      return;
+    }
+    const { type } = message;
+    if (type === MessageTypes.GQL_STOP) {
+      nextTick(() => {
+        this.processQueue(connectionContext, opId);
+      });
+      this.unsubscribe(connectionContext, opId);
+      return;
+    }
+    // START
+    this.processStart(connectionContext, operationContext, opId, message).catch(error => {
+      operationContext.processingStart = false;
+      nextTick(() => {
+        this.processQueue(connectionContext, opId);
+      });
+      this.sendError(connectionContext, opId, { message: error.message });
+    });
+  }
+
+  private queueMessage(connectionContext: ConnectionContext, opId: string, message: OperationMessage) {
+    let operationContext = connectionContext.operations.get(opId);
+    if (!operationContext) {
+      operationContext = {
+        messageQueue: [],
+        processingStart: false,
+      };
+      connectionContext.operations.set(opId, operationContext);
+    }
+    operationContext.messageQueue.push(message);
+    this.processQueue(connectionContext, opId);
+  }
+
   private onMessage(connectionContext: ConnectionContext) {
     return (message: any) => {
       let parsedMessage: OperationMessage;
       try {
         parsedMessage = parseLegacyProtocolMessage(connectionContext, JSON.parse(message));
       } catch (e) {
-        this.sendError(connectionContext, null, { message: e.message }, MessageTypes.GQL_CONNECTION_ERROR);
+        this.sendError(connectionContext, undefined, { message: e.message }, MessageTypes.GQL_CONNECTION_ERROR);
         return;
       }
 
-      const opId = parsedMessage.id;
-      switch (parsedMessage.type) {
+      const { id: opId, payload, type } = parsedMessage;
+      switch (type) {
         case MessageTypes.GQL_CONNECTION_INIT:
-          if (this.onConnect) {
+          const onConnect = this.onConnect;
+          if (onConnect) {
             connectionContext.initPromise = new Promise((resolve, reject) => {
```

Comment:
Any reason not to async/await-ify this?

Reply:
That would involve another Promise wrapper or converting onMessage to async. As this wasn't related to the root problem I'm deferring that for now.